### ISSUE TITLE
stale rank caching

### ISSUE THREAD

**Issue by rdm** [jsoftware/jsource#142]
https://mlochbaum.github.io/BQN/commentary/why.html#named-functions at the bottom illustrates an old bug in J:

```
   g=:*
   f=:g
   g=:|:
g b.0
_ 1 _
   f b.0
0 0 0
```

Or, for a more amusing illustration of this issue:

```
   g=:*
   f=:g
   g=: |: * *@#@$
   f i.3 3
0 3 6
1 4 7
2 5 8
   f f i.3 3
0 1 2
3 4 5
6 7 8
   f@f i.3 3
0 0 0
0 0 0
0 0 0
```

**Comment by moon-chilled** [comment]
I do not think this is a bug; you might rather call it a quirk.  Tacit definitions are expected to be closed, but exactly what they close over might be other than what you would expect.  Currently, verb _definitions_ are late-bound, but verb _ranks_ are early-bound.  I don't see a way around this.  To see why, consider the following:

```
c=: {{
 echo 'hi'
 if. {.u b.0 do. u@:v else. u&:v end. }}
g=: -
a=: f c g
f=: +
```

In order to get the behaviour you desire, it would be necessary to re-apply c (including multiple evaluation of any embedded side effects) here, which is clearly undesirable.

By the by, discussion of f. there is also not quite right; better to use u./v. in modifiers.

**Comment by rdm** [comment]
I see `{{ if. {. u b. 0 do. ... end. }}` as a different issue:

Adverbs and Conjunctions evaluate "early". Changing a part of a definition after it was built by a conjunction should be clearly within the realm of the programmer's efforts.

But if rank is not a part of the verb's definition, that's at the very least something that should be clearly documented but reading through the historical treatment of rank and copula, neither mention anything which would suggest this issue. And, yet this is a very old issue. (So, it seems to me that this behavior was unintended and overlooked -- a "bug".)

Put different, this is very difficult to describe. In the above examples, where f has the definition g, talking about why f has the definition g but the rank of f is different from the rank of g. It's inspectable, but it's difficult to talk about.

Also, in my experience, when it's easier to change the code to make it easy to describe than it is to fix the documentation to clearly describe the code, it's best to fix the code.

**Comment by moon-chilled** [comment]
Perhaps my point will be clearer if I make it differently.  Maybe consider this example instead:

```
g=. -
a=. f@g
b=. g
g=. |.
```

Under your proposed semantics, a will 'cache' the rank of g, but b will not.  Why this inconsistency?

Notionally, a combinator is a higher-order function which does not 'look inside' of its function arguments, but only applies them, variably.  While modifiers in j are not required to be combinatoric, they often (almost always?) are so in practice.  This combinatory behaviour is what enables coherent abstraction and makes late binding work.

However, if we look at the built-in modifiers, we see one particular category of counter-examples: the rank-preserving conjunctions—: :. & &. @ (forgive me if I've missed any; I've intentionally discounted u"v).  Though these do 'peek inside' of their argument verbs, the only mote of early binding that they do is to find those verbs' ranks.  In keeping with these (and considering the fundamental role these modifiers play) I find it most coherent for verb rank to _always_ be early-bound.

> But if rank is not a part of the verb's definition

It _is_ part of the verb's definition!  Above, I define b; that definition for b includes a rank.  Why when I redefine g should b change?  I have not changed the definition of b; only of g.

> But reading through the historical treatment of rank and copula, neither mention anything which would suggest this issue. And yet, this is a very old issue. (So, it seems to me that this behavior was unintended and overlooked -- a "bug".)

It is often difficult to predict what the consequences of some design choice will be.  FWIW, my clean-slate solution is to get rid of verb rank entirely—while useful, it does not bear its weight, especially given it is so heavy as to cause troubles like these.  (This, along with changes to the way name lookup is done, would resolve all the problems brought up by your link while sacrificing none of the advantages of late binding.)  But this hand is dealt, and I find the current behaviour the most coherent and consistent one.

> in my experience, when it's easier to change the code to make it easy to describe than it is to fix the documentation to clearly describe the code, it's best to fix the code

I do not think the implementation should be anything more than a distant concern, while writing the specification.  (IOW, it doesn't matter how easy it is to change the code, so long as it can be done at all.)

**Comment by rdm** [comment]
Except, I did not propose any semantics.

I shall do so now:

I propose that when the definition of a name is changed, that all definitions which tacitly include that name be marked as stale, so that the new definition is applied consistently in all contexts.

**Comment by moon-chilled** [comment]
> I did not propose any semantics

Fair enough.  I assumed them, apparently incorrectly.  But I find your proposal even more complicated to explain.

What is this notion of 'staleness'?  It would seem to me that your proposal would have c applied multiple times in my original example.  If so, why?  If not, why do you privilege some forms of 'caching' (I do not think that is a sensible use of the word, but w/e) over others?

**Comment by rdm** [comment]
It's basically just cache invalidation.

**Comment by moon-chilled** [comment]
1. What sorts of definitions are eligible for caching?

2. What happens when a 'cached' definition is invalidated?

**Comment by rdm** [comment]
The current evidence I see suggests that the rank of definition which was associated with a name is cached in at least some tacit definitions. (But I am not completely sure what you are asking here: were you suggesting that you had not noticed this issue, or were you asking me whether other things are cached in the context of tacit definitions? If the latter, I would note that future implementations might be caching information which is not currently being cached.)

When a cached tacit definition is invalidated, this would mean that the names used in that tacit definition must be dereferenced, to get the details of their definitions (some of which might then be cached for future evaluations of that tacit definition) during or before the evaluation of that tacit definition.

**Comment by HenryHRich** [comment]
Reasonable.

It would apply only when non-nouns are assigned, which is seldom. It 
could be implemented by traversing every locale to recalculate the rank 
of every verb whenever any non-noun is assigned.

[Faster would be to chain verbs off of the verbs that they depend on, 
but that would swell the verb definition to another cacheline.]

I vote against the proposal.

hhr

**Comment by rdm** [comment]
I guess I was thinking of chaining verbs off of verbs they depend on.

I don't think cachelines very much matter here, because this chaining data structure would only be referenced when verb definitions were being updated. (Or were you recommending against the proposal for some other reason?)

**Comment by moon-chilled** [comment]
> But I am not completely sure what you are asking here: were you suggesting that you had not noticed this issue, or were you asking me whether other things are cached in the context of tacit definitions? If the latter, I would note that future implementations might be caching information which is not currently being cached

I am aware of the following issue: if one reassigns a verb which was previously referred to by name in a tacit definition, and if this changes the rank of the verb, the resultant behaviour may be nonintuitive.  I think this is also the issue you were referring to.

I find the use of 'caching' somewhat strange here, because it seems to prescribe a particular implementation strategy, whereas what I am interested in is the semantics and the behaviour, not the particulars of the implementation; but it's the word you've used, so I'm rolling with it.

I want to know exactly what the behaviour is that you're after, because it's not quite clear to me.  Presumably, in my second example, when g is reassigned, you would like the rank of b to change with it; yes?  Would you also like the rank of a to change?  Would you like for c to be re-applied in the first example?

(My opinion is that if the answer to all three of these is not the same, then the policy is inconsistent, and that an answer of yes to the third question is too horrible to countenance, so the answer to all three should be no.)

**Comment by moon-chilled** [comment]
> It would apply only when non-nouns are assigned, which is seldom

And at that, only when verb ranks change, which should be even rarer.

**Comment by rdm** [comment]
> > But I am not completely sure what you are asking here: were you suggesting that you had not noticed this issue, or were you asking me whether other things are cached in the context of tacit definitions? If the latter, I would note that future implementations might be caching information which is not currently being cached
> 
> I am aware of the following issue: if one reassigns a verb which was previously referred to by name in a tacit definition, and if this changes the rank of the verb, the resultant behaviour may be nonintuitive. I think this is also the issue you were referring to.
> 
> I find the use of 'caching' somewhat strange here, because it seems to prescribe a particular implementation strategy, whereas what I am interested in is the semantics and the behaviour, not the particulars of the implementation; but it's the word you've used, so I'm rolling with it.

What term would you describe this behavior which is less strange than "caching"?


> I want to know exactly what the behaviour is that you're after, because it's not quite clear to me. Presumably, in my second example, when g is reassigned, you would like the rank of b to change with it; yes? Would you also like the rank of a to change? Would you like for c to be re-applied in the first example?

I am puzzled here. On the one hand, you describe the current behavior as "perhaps nonintuitive". And you also seem to reject my summary description of the current behavior as stale cache.

So, I'll try phrasing my point of view using different words:

J implemented "late binding" for verbs, so that verb definitions could be supplied after the verb names were used in definitions. I think that the rank of the verb (which is a part of the definition) should be respected when a definition is associated with a late bound name.

When this should have further effects (when the definition itself uses a combinator which is sensitive to the rank of the verb), then yes the updated verb rank should still have this effect.

> (My opinion is that if the answer to all three of these is not the same, then the policy is inconsistent, and that an answer of yes to the third question is too horrible to countenance, so the answer to all three should be no.)

If by "second example" you meant

```
g=. -
a=. f@g
b=. g
g=. |.
```

I do not understand what is "too horrible to countenance". Is that something you could explain?

**Comment by moon-chilled** [comment]
There were three questions:

1. in my second example, when g is reassigned, you would like the rank of b to change with it; yes?

2. Would you also like the rank of a to change? [also in the second example]

3. Would you like for c to be re-applied in the first example?

The first example is the one which includes c, a user-defined conjunction with side effects.  It is question #3 which is too horrible to countenance an answer of 'yes' to, because users do not like it when side effects are unexpectedly performed ('spooky action at a distance').

**Comment by moon-chilled** [comment]
> What term would you describe this behavior which is less strange than "caching"?

I say that verb rank is bound early (for consistency with a number of primitive conjunctions which are closed over the ranks of their operands).

**Comment by rdm** [comment]
> There were three questions:
> 
> 1. in my second example, when g is reassigned, you would like the rank of b to change with it; yes?
> 2. Would you also like the rank of a to change? [also in the second example]
> 3. Would you like for c to be re-applied in the first example?
> 
> The first example is the one which includes c, a user-defined conjunction with side effects. It is question #3 which is too horrible to countenance an answer of 'yes' to, because users do not like it when side effects are unexpectedly performed ('spooky action at a distance').

Ok, now I think you are talking about

```
c=: {{
 echo 'hi'
 if. {.u b.0 do. u@:v else. u&:v end. }}
g=: -
a=: f c g
f=: +
```

(Personally, I thought of this as either the third example in the thread -- since i had provided two different examples, or the first example that you had provided. But I now think it's clear that you had thought of as both of my examples as being a single example. That said... I could still be incorrect in some way. Please let me know if I have still managed to incorrectly identify your concept of the second example.)

But here, no, I see no reason for a to change if c changes.

Here, the definition of a is  `f&:g` and c is not a part of that definition. (We know that c was used in constructing that definition, but c is a conjunction which is necessarily used with "early binding".)

**Comment by moon-chilled** [comment]
What you reproduced is what I meant to refer to as my first example.

_Notionally_, a primitive conjunction such as @ is no different from c.  @ itself is not 'part' of f@g; it is applied, and then returns a new verb.  So, again: if we have a=. f@g, and we redefine g to have a different rank, should we expect the rank of a to change?

**Comment by moon-chilled** [comment]
here's another, more concrete problem: how do you deal with cycles?

```
a=. -
b=. a
a=. b
```

**Comment by rdm** [comment]
> What you reproduced is what I meant to refer to as my first example.
> 
> _Notionally_, a primitive conjunction such as @ is no different from c. @ itself is not 'part' of f@g; it is applied, and then returns a new verb. So, again: if we have a=. f@g, and we redefine g to have a different rank, should we expect the rank of a to change?

J provides two approaches for explicit conjunctions. If the user wanted to have c remain a part of the definition, the definition of c would reference y (and, optionally: x). 

> here's another, more concrete problem: how do you deal with cycles?
> 
> ```
> a=. -
> b=. a
> a=. b
> ```

There's several possible approaches here.

a=. -   NB. a gets rank 0
b=. a  NB. b gets rank 0
a=. b  NB. a gets rank 0 -- no change, no need to invalidate b's rank

-------------

a=. -  NB. a marked stale, rank will be resolved at runtime
b=. a  NB b marked stale, rank will be resolved at runtime
a=. b  NB. a marked stale, rank will be resolved at runtime, since a is already stale no need for further action

-------------

a=. - NB. a's rank set to 0 since it has no name dependencies
b=. a NB. b's rank marked stale, will be resolved at runtime
a=. b NB. a marked stale and references marked stale
NB. since b has already been marked stale, the marking process does not progress past b

-------------

a=. - NB. a's rank set to zero, since a has no named dependencies
b=. a NB. b's rank set to zero, since a is not stale
a=. b NB. a's rank does not change, no need to mark it stale

**Comment by moon-chilled** [comment]
> If the user wanted to have c remain a part of the definition, the definition of c would reference y (and, optionally: x).

Yes, but this is obviously incompatible with explicitly setting rank (as is necessary for something like @).

**Comment by rdm** [comment]
> > If the user wanted to have c remain a part of the definition, the definition of c would reference y (and, optionally: x).
> 
> Yes, but this is obviously incompatible with explicitly setting rank (as is necessary for something like @).

If you are saying that the interpreter uses functional details which are not completely exposed to the programmer using the interpreter, that would be correct (and: unavoidable).

If you are saying that the interpreter should make those functional details available to the, I don't think that that follows.

If you are saying something entirely different, I think I do not understand what you are saying (perhaps you could elaborate, if this is the case?).
